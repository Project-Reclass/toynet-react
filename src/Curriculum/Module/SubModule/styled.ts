/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import styled from '@emotion/styled';
import { Box, Link, Text } from '@chakra-ui/react';

interface ModuleNameProps {
  locked: boolean;
  hoverColor?: string;
}

export const ModuleName = styled(Link)`
  color: ${({ locked }: ModuleNameProps) => locked ? 'grey' : 'white'};
  display: flex;
  cursor: pointer;
  justify-content: space-between;
  align-items: center;
  a {
    -webkit-user-select: none;
    user-select: none;
    svg {
      width: 1.5em;
      height: 1.5em;
    }
  }
  :hover {
    color: ${({ hoverColor }: ModuleNameProps) => hoverColor || 'white'};
  }
`;

interface SIBoxProps {
  type: 'lock' | 'check';
}

export const StatusIconBox = styled(Box)`
  background-color: ${({type}: SIBoxProps) => type !== 'check' ? 'white' : 'rgb(79, 149, 26)'};
  border-radius: 3px;
  width: 1.5rem;
  height: 1.5rem;
`;

export const SubmoduleDescription = styled(Text)`
  font-size: 14px;
  width: 80%;
  margin-left: 3rem;
`;