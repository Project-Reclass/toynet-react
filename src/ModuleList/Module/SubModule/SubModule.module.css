.dropdownItem {
  width: 75%;
  font-size: 1rem;
  box-shadow:  0pt 0px 3px rgba(0, 0, 0, 0.3);
  color: white;
  border-radius: 3px;
  background-color: rgba(255, 255, 255, 0.05);
  display: grid;
  grid-template-columns: 1fr 1fr;

  padding: 0.5rem 0;
  margin: 0.229rem;
}

.dropdownItem > span {
  margin: auto 0.5rem;
  text-align: center;
}
