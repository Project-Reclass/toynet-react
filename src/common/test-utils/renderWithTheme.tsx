import React, { FC } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '@chakra-ui/core/dist';
import renderer from 'react-test-renderer';

const Wrapper: FC = ({ children }) => (
  <ThemeProvider>{children}</ThemeProvider>
);

export const renderWithTheme = (ui: any) => {
  return render(ui, { wrapper: Wrapper });
};

export const renderTreeWithTheme = (ui: any) => {
  return renderer.create(<Wrapper>{ui}</Wrapper>);
};
