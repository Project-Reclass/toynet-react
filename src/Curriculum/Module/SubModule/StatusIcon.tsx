import React from 'react';
import { Stack, Icon } from '@chakra-ui/core';

import { StatusIconBox } from './styled';

interface SIProps {
  status: 'locked' | 'unlocked' | 'done'
}

const StatusIcon = ({ status }: SIProps) => (
  <Stack>
    <StatusIconBox type={status !== 'done' ? 'lock' : 'check'}>
      <Icon
        name={status === 'done' ? 'check' :
          status === 'locked' ? 'lock' : 'unlock'
        }
        margin='auto'
        marginTop='0.2rem'
        display='block'
        color={status !== 'done' ? 'grey' : 'white'}
      />
    </StatusIconBox>
  </Stack>
);


export default StatusIcon;
