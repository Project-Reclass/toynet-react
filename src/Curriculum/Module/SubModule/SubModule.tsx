/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import { FC } from 'react';
import { Divider, Flex, Link, Stack, Text, Collapse, Tooltip } from '@chakra-ui/react';
import { StarIcon } from '@chakra-ui/icons';
import { BsPlayCircle } from 'react-icons/bs';
import { SubModuleIntf } from 'src/common/types/curriculum';

import { ModuleName } from './styled';
import { useSessionStorage } from 'src/common/hooks/useSessionStorage';
import { createLink } from 'src/common/utils';
import { SubmoduleDescription } from './styled';

interface Props extends SubModuleIntf {
  moduleId: number;
  index: number;
  count: number;
}


const capitalize = (s: string): string =>
  `${s[0].toUpperCase()}${s.toLowerCase().slice(1)}`;

export const SubModule: FC<Props> = (
  {
    name,
    id,
    moduleId,
    type,
    index,
    count,
    introduction,
  },
) => {
  const [isOpen, setOpen] =
    useSessionStorage<boolean>(`submodule-${moduleId}-${id}-${index}`, true,
      value => JSON.parse(value));

  return (
    <Flex my='1'>
      <StarIcon
        w='1.5rem'
        h='1.5rem'
        color='yellow.400'
      />
      <Stack spacing={2} width='100%' marginLeft='1.5rem'>
        <Flex justifyContent='space-between'>
          <ModuleName locked={false}>
            <Tooltip
              hasArrow
              label={isOpen ? 'Show less' : 'Expand submodule'}
            >
              <Text onClick={() => setOpen(open => !open)}>
                {`${capitalize(type.toString())}: ${name}`}
              </Text>
            </Tooltip>
          </ModuleName>
          <ModuleName locked={false} hoverColor='rgba(84,143,155)'>
            <Link href={createLink({ moduleId, type, id })}>
              <BsPlayCircle data-testid='submodule_play_icon' />
            </Link>
          </ModuleName>
        </Flex>
        <Collapse in={isOpen} color='white'>
          <SubmoduleDescription>
            {introduction}
          </SubmoduleDescription>
        </Collapse>
        {index !== count -1 && <Divider />}
      </Stack>
    </Flex>
  );
};


export default SubModule;