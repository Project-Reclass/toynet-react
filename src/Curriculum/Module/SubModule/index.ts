import SubModule from './SubModule';
export default SubModule;