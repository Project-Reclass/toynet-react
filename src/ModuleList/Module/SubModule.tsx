import React, { FC } from 'react';
import styled from '@emotion/styled';

import { ModuleInterface } from './Module';

const DropdownItem = styled.div`
  width: 75%;
  font-size: 1rem;
  box-shadow:  0pt 0px 3px rgba(0, 0, 0, 0.3);
  color: white;
  border-radius: 3px;
  background-color: rgba(255, 255, 255, 0.05);
  display: grid;
  grid-template-columns: 1fr 1fr;

  padding: 0.5rem 0;
  margin: 0.229rem;

  span {
    margin: auto 0.5rem;
    text-align: center;
    user-select: none;
  }
`;

const SubModule: FC<ModuleInterface> = ({ title, progress, id, moduleId, type }) => (
  <DropdownItem>
    <span>
      <a href={`/module/${moduleId}/${type.toString()}/${id}`}>
        {title}
      </a>
    </span>
    <span>
      {/* <ProgressBar label={`${progress}%`} now={progress} /> */}
    </span>
  </DropdownItem>
);

export default SubModule;