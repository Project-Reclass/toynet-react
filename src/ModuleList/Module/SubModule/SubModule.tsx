import React, { FC } from 'react';
import { ProgressBar } from 'react-bootstrap';

import { ModuleInterface } from '../Module';

import styles from './SubModule.module.css';

const SubModule: FC<ModuleInterface> = ({ title, progress, id, moduleId, type }) => (
  <div className={styles.dropdownItem}>
    <span className={styles.title}>
      <a href={`/module/${moduleId}/${type.toString()}/${id}`}>
        {title}
      </a>
    </span>
    <span className={styles.progress}>
      <ProgressBar label={`${progress}%`} now={progress} />
    </span>
  </div>
);

export default SubModule;