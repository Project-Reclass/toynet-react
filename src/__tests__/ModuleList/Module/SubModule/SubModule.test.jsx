import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react'

import SubModule from 'src/Curriculum/Module/SubModule/SubModule';
import { renderTreeWithTheme, renderWithTheme } from 'src/common/test-utils/renderWithTheme';

const defaultProps = {
  title: 'Article',
  progress: 99,
  id: 1,
  moduleId: 2,
  type: 'article',
}

const valuesProps = {
  title: 'Article',
  progress: 99,
  id: 1,
  moduleId: 2,
  type: 'value',
}

describe('The sub modules', () => {
  it('should render and match previous snapshots', () => {
    const tree = renderTreeWithTheme(
      <SubModule {...defaultProps} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should create a link to the submodule page', () => {
    const { type, id, title } = defaultProps;
    const { getByText } = renderWithTheme(<SubModule {...defaultProps} />);
    const link = getByText(title);

    expect(link.getAttribute('href')).toBe(`/module/0/${type.toString()}/${id}`)
  });
  it('should create a link to a values page', () => {
    const { type, id, title } = valuesProps;
    const { getByText } = renderWithTheme(<SubModule {...valuesProps} />);
    const link = getByText(title);

    expect(link.getAttribute('href')).toBe(`/${type.toString()}/${id}`)
  })
});