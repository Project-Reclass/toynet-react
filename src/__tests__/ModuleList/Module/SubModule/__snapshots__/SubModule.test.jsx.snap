// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The sub modules should render and match previous snapshots 1`] = `
<div
  className="css-4ngnmx"
>
  <span>
    <a
      href="/module/2/article/1"
    >
      Article
    </a>
  </span>
  <span />
</div>
`;
